File: algorithm/003-longest_substring_without_repeating_characters/swift/solution.swift
class Solution {
  func lengthOfLongestSubstring(_ s: String) -> Int {
        var alphabet = [Character: Int]()
        var longest = 0
        var i = 0
        var j = 0
        
        for c in s {
            if alphabet[c] != nil {
                longest = max(longest, j - i)
                while i < alphabet[c]! {
                    alphabet[s[s.index(s.startIndex, offsetBy: i)]] = 0
                    i += 1
                }
            }
            j += 1
            alphabet[c] = j
        }
        
        return max(longest, j - i)
    }
}

File: algorithm/003-longest_substring_without_repeating_characters/ruby/solution.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    alphabet = {}
    longest = 0
    i = 0
    j = 0

    s.each_char{
      |c|
      if alphabet[c]
        longest = [longest, j - i].max
        while i < alphabet[c]
          alphabet[s[i]] = 0
          i += 1
        end
      end

      j += 1
      alphabet[c] = j
    }

    return [longest, j - i].max
end